hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go

```diff
@@ -645,6 +647,59 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	createOrUpdate := r.createOrUpdate(req)
 
+	// Convert old configuration fields to new fields
+
+	if hcluster.Spec.Configuration != nil && len(hcluster.Spec.Configuration.Items) > 0 {
+		originalSpec := hcluster.Spec.DeepCopy()
+		gconfig, err := globalconfig.ParseGlobalConfig(ctx, hcluster.Spec.Configuration)
+		if err != nil {
+			// This should never happen because at this point, the global configuration
+			// should be valid
+			return ctrl.Result{}, err
+		}
+
+		// Only copy over config from the raw extension if the field is not already populated.
+		// Once populated, the field takes precedence and a conflicting raw extension config
+		// results in an invalid configuration field.
+
+		if gconfig.APIServer != nil && hcluster.Spec.Configuration.APIServer == nil {
```

Comment:
nit: do we really need to include the first checks in the ifs `gconfig.APIServer != nil` or is it redundant?

Reply:
Yes we do need a check because it will be nil if you haven't included an APIServer resource in your list of extensions. In which case, there's nothing to set for the APIServer field.

Reply:
>In which case, there's nothing to set for the APIServer field.
in which case we would just set nil (without the check), right? I was curious if there was a reason to make nil a special case and have check vs just reconcile it, being nil or not.

Reply:
they are different things... so we can't just set one to the other. In the raw extension case, it's the whole resource because it needs to be. In the new field case, it's only the spec.